File: app/src/main/java/com/aar/app/wsp/features/gameover/GameOverActivity.java
package com.aar.app.wsp.features.gameover;

import androidx.lifecycle.ViewModelProviders;
import android.content.Intent;
import androidx.core.app.NavUtils;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.aar.app.wsp.R;
import com.aar.app.wsp.features.ViewModelFactory;
import com.aar.app.wsp.WordSearchApp;
import com.aar.app.wsp.commons.DurationFormatter;
import com.aar.app.wsp.features.gameplay.GamePlayActivity;
import com.aar.app.wsp.model.GameData;
import com.aar.app.wsp.features.FullscreenActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class GameOverActivity extends FullscreenActivity {
    public static final String EXTRA_GAME_ROUND_ID =
            "com.paperplanes.wsp.presentation.ui.activity.GameOverActivity";

    @Inject
    ViewModelFactory mViewModelFactory;

    @BindView(R.id.textCongrat) TextView mTextCongrat;
    @BindView(R.id.game_stat_text) TextView mTextGameStat;

    private GameOverViewModel mViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_over);

        ButterKnife.bind(this);
        ((WordSearchApp) getApplication()).getAppComponent().inject(this);

        mViewModel = ViewModelProviders.of(this, mViewModelFactory).get(GameOverViewModel.class);
        mViewModel.getOnGameDataLoaded().observe(this, this::showGameStat);
        mViewModel.getOnGameDataReset().observe(this, gameDataId -> {
            Intent i = new Intent(GameOverActivity.this, GamePlayActivity.class);
            i.putExtra(GamePlayActivity.EXTRA_GAME_DATA_ID, gameDataId);
            startActivity(i);
            finish();
        });

        if (getIntent().getExtras() != null) {
            int gid = getIntent().getExtras().getInt(EXTRA_GAME_ROUND_ID);
            mViewModel.loadData(gid);
        }
    }

    @OnClick(R.id.main_menu_btn)
    public void onMainMenuClick() {
        onBackPressed();
    }

    @OnClick(R.id.btnReplay)
    public void onReplayClick() {
        mViewModel.resetCurrentGameData();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        goToMainMenu();
    }

    private void goToMainMenu() {
        if (getPreferences().deleteAfterFinish()) {
            mViewModel.deleteGameRound();
        }
        NavUtils.navigateUpTo(this, new Intent());
        finish();
    }

    public void showGameStat(GameData gd) {
        if (gd.isGameOver()) {
            mTextCongrat.setText(R.string.lbl_game_over);
            mTextGameStat.setVisibility(View.GONE);
        } else {
            String strGridSize = gd.getGrid().getRowCount() + " x " + gd.getGrid().getColCount();

            String str = getString(R.string.finish_text);
            str = str.replaceAll(":gridSize", strGridSize);
            str = str.replaceAll(":uwCount", String.valueOf(gd.getUsedWords().size()));
            str = str.replaceAll(":duration", DurationFormatter.fromInteger(gd.getDuration()));

            mTextCongrat.setText(R.string.congratulations);
            mTextGameStat.setVisibility(View.VISIBLE);
            mTextGameStat.setText(str);
        }
    }
}

File: app/src/main/java/com/aar/app/wsp/features/gameover/GameOverActivity.kt
package com.aar.app.wsp.features.gameover

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.core.app.NavUtils
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.observe
import com.aar.app.wsp.R
import com.aar.app.wsp.WordSearchApp
import com.aar.app.wsp.commons.DurationFormatter.fromInteger
import com.aar.app.wsp.databinding.ActivityGameOverBinding
import com.aar.app.wsp.features.FullscreenActivity
import com.aar.app.wsp.features.gameplay.GamePlayActivity
import com.aar.app.wsp.model.GameData
import kotlinx.coroutines.launch
import javax.inject.Inject

class GameOverActivity : FullscreenActivity() {
    @Inject
    lateinit var mViewModelFactory: ViewModelProvider.Factory

    private val viewModel: GameOverViewModel by viewModels { mViewModelFactory }

    private lateinit var binding: ActivityGameOverBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityGameOverBinding.inflate(layoutInflater)
        setContentView(binding.root)

        (application as WordSearchApp).appComponent.inject(this)

        setupViews()
        setupViewModel()
        loadGameData()
    }

    private fun setupViews() {
        binding.mainMenuBtn.setOnClickListener {
            onBackPressed()
        }
        binding.btnReplay.setOnClickListener {
            resetCurrentGameData()
        }
    }

    private fun setupViewModel() {
        viewModel.onGameDataLoaded.observe(this) { gameData ->
            gameData?.let { showGameStat(it) }
        }
        viewModel.onGameDataReset.observe(this) { gameDataId ->
            goToGamePlay(gameDataId)
        }
    }

    private fun goToGamePlay(gameDataId: Int) {
        val intent = Intent(this@GameOverActivity, GamePlayActivity::class.java)
        intent.putExtra(GamePlayActivity.EXTRA_GAME_DATA_ID, gameDataId)
        startActivity(intent)
        finish()
    }

    private fun resetCurrentGameData() {
        lifecycleScope.launch {
            viewModel.resetCurrentGameData()
        }
    }

    private fun loadGameData() {
        intent.extras?.getInt(EXTRA_GAME_ROUND_ID)?.let { gameId ->
            lifecycleScope.launch { viewModel.loadData(gameId) }
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        goToMainMenu()
    }

    private fun goToMainMenu() {
        if (preferences.deleteAfterFinish()) {
            lifecycleScope.launch { viewModel.deleteGameRound() }
        }
        NavUtils.navigateUpTo(this, Intent())
        finish()
    }

    private fun showGameStat(gd: GameData) {
        if (gd.isGameOver) {
            binding.textCongrat.setText(R.string.lbl_game_over)
            binding.gameStatText.visibility = View.GONE
        } else {
            val strGridSize = gd.grid!!.rowCount.toString() + " x " + gd.grid!!.colCount
            var str = getString(R.string.finish_text)
            str = str.replace(":gridSize".toRegex(), strGridSize)
            str = str.replace(":uwCount".toRegex(), gd.usedWords.size.toString())
            str = str.replace(":duration".toRegex(), fromInteger(gd.duration))
            binding.textCongrat.setText(R.string.congratulations)
            binding.gameStatText.visibility = View.VISIBLE
            binding.gameStatText.text = str
        }
    }

    companion object {
        const val EXTRA_GAME_ROUND_ID = "com.paperplanes.wsp.presentation.ui.activity.GameOverActivity"
    }
}